### PR TITLE
Introduction of BrowserStack testing framework and initial test cases

### PR DESCRIPTION
**Pull request by jGonzalez-CCPO**
BrowserStack testing framework allows the execution of Selenium-based tests written in Python.  Tests can be run with any of the target browsers Chrome, Edge, and IE11.  See `uitests/framework/README.md`

### AT-6112  Define common elements for testing framework in "master" page object to provide modularity to the test cases
- Master Page
- TO Page
- Portfolio Page
- Settings Page
- Login Page
- Application Page
- Reports

### AT-6113  Initial test cases that use the above framework
- Login Scenario
- Create Portfolio

### AT-6115  Implement test cases
- Task Order Creation
- Application Creation (Includes Environment Information)

Reports to validate:
- Task Order Information
- Application Information
- Environment Information

Tickets: AT-6112, AT-6113, AT-6115

### CODE REVIEW

**Comment by jGonzalez-CCPO** [comment]
Pending changes

**Comment by jGonzalez-CCPO** [comment]
pending changes

File: pyproject.toml [hunk 1]
```diff
@@ -38,6 +38,7 @@ werkzeug = "*"
 email-validator = '*'
 python3-saml = "^1.9.0"
 Flask-DebugToolbar = "^0.11.0"
+selenium = "^3.141.0"
```
Comment (jasonburkert-ccpo) [inline pyproject.toml:41]: This should probably be under `dev-dependencies`.   I would expect there should be a corresponding targeted change to file `poetry.lock` with this new dependency.

File: .gitignore [hunk 2]
```diff
@@ -116,3 +116,5 @@ js/test_templates
 
 # npm lock file; using yarn
 package-lock.json
+
+dump.rdb
```
Comment (jasonburkert-ccpo) [inline .gitignore:120]: Can we add a comment about the file being ignored?

File: pyproject.toml (hunk 1 above)
Comment (jGonzalez-CCPO) [reply pyproject.toml:41]: it was there already in fack it was 2 times

Comment (jGonzalez-CCPO) [reply pyproject.toml]: selenium = "^3.141.0" 
selenium = "*"

File: uitests/TestFramework/Driver_Notes/LICENSE:Zone.Identifier [hunk 3]
```diff
@@ -0,0 +1,3 @@
+[ZoneTransfer]
+ZoneId=3
+ReferrerUrl=C:\Users\EdwardHoward(ASG)\Downloads\edgedriver_win64 (3).zip
```
Comment (jasonburkert-ccpo) [inline uitests/TestFramework/Driver_Notes/LICENSE:Zone.Identifier]: Do we need this ReferrerUrl in here?  Probably shouldn't have this user-specific path in the repo.

Comment (jasonburkert-ccpo) [reply uitests/TestFramework/Driver_Notes/LICENSE:Zone.Identifier]: This applied to the other files ending in "Zone.Identifier" as well.

File: uitests/TestFramework/pageObjects/CommonMethods.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+from selenium.webdriver.support.ui import Select
+from selenium.webdriver.common.keys import Keys
+from selenium.webdriver.common.by import By
+from selenium import webdriver
+
+
+class Jedi_Common_Methods:
```
Comment (jasonburkert-ccpo) [inline uitests/TestFramework/pageObjects/CommonMethods.py]: The fate of the DoD JEDI contract is uncertain.  We should avoid using this term.

File: uitests/TestFramework/Driver_Notes/LICENSE:Zone.Identifier (hunk 3 above)
Comment (kylelaker-ccpo) [reply uitests/TestFramework/Driver_Notes/LICENSE:Zone.Identifier]: A `Zone.Identifier` should never be committed. These are artifacts of copying from Windows host to WSL. They can all be removed (both here and locally).

File: uitests/TestFramework/testCases/conftest.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    {
+        'os_version': '10',
+        'os': 'Windows',
+        'browser': 'chrome',  # Edge  Chrome  IE
+        'browser_version': '89.0',  # 88.0  89.0    11.0
+        'resolution': '1920x1080',
+        'browserstack.sendKeys': 'true',
+        'browserstack.debug': 'true',
+        'build': 'parallel test #4'  # Your tests will be organized within this build
+    }]
```
Comment (jasonburkert-ccpo) [inline uitests/TestFramework/testCases/conftest.py]: There's far too much duplication of the same code here.  Most of the content of these python dicts is repetitive.  Please follow the DRY principal.  There's 60+ lines here, aim for less than half that.

File: uitests/TestFramework/testCases/conftest.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    elif browser == 'all':
+        driver = webdriver.Remote(
+            command_executor=address,
+            desired_capabilities=allBrowsers)
+        print("Launching IE 11 Browser.....")
+    # elif browser == 'firefox':
+    #     driver = webdriver.Remote(
+    #         command_executor=address,
+    #         desired_capabilities=firefox)
+    #     print("Launching Firefox Browser.....")
```
Comment (jasonburkert-ccpo) [inline uitests/TestFramework/testCases/conftest.py]: Let's remove this.

File: uitests/TestFramework/testCases/test_create_app.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+from selenium.common.exceptions import TimeoutException
+from selenium.webdriver.support.wait import WebDriverWait
+from utilities.readProperties import ReadConfig
+from pageObjects.ApplicationPage import CreateApplicationPages
+
+time_now = datetime.datetime.now().strftime('%m%d%Y%H%M%S')
+
+
+class Test_004_Create_Application:
+    url2 = ReadConfig.getLoginLocalURL()
```
Comment (jasonburkert-ccpo) [inline uitests/TestFramework/testCases/test_create_app.py]: Why _url2_?  Is there a _url1_?  Perhaps __url_ to indicate it is only used internally?  Reference: https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

File: uitests/TestFramework/utilities/readProperties.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+config = configparser.RawConfigParser()
+config.read(".//Configurations//config.ini")
+
+
+class ReadConfig:
+    @staticmethod
+    def getApplicationURL():
+        url = os.getenv("baseURL")
+        return url
```
Comment (jasonburkert-ccpo) [inline uitests/TestFramework/utilities/readProperties.py]: Refactor to simply `return os.getenv("baseURL")`.

Apply same to other _get_ methods.

File: uitests/TestFramework/Driver_Notes/LICENSE:Zone.Identifier (hunk 3 above)
Comment (jGonzalez-CCPO) [reply uitests/TestFramework/Driver_Notes/LICENSE:Zone.Identifier]: uitests/TestFramework/Driver_Notes/LICENSE:Zone.Identifier 

this was take care in a latter commit

**Comment by kylelaker-ccpo** [comment]
`static/img/test.pdf` should not be at that location. Files under that path are always accessible on a deployed application. For example, this would be available under `https://<namespace.atat.dev/static/img/test.pdf` just as `static/img/minus.svg` is available at `https://<namespace>.atat.dev/static/img/minus.svg`.

Files to support tests should be stored in the `tests/` or `uitests/` directory as appropriate.

File: uitests/framework/README.md [hunk 9]
```diff
@@ -99,7 +96,7 @@ BrowserStack that can change on time so verify with your provider. Ex.:
 
 ```
 # Running a test with the BrowserStack's chrome drivers in parallel. 
- (.venv) $ pytest -s -v -n=2 uitests/framework/test_cases/test_login.py --browser chrome
+ (.venv) $ pytest -n=2 -s -v -m smoke uitests/* --browser chrome
```
Comment (jasonburkert-ccpo) [inline uitests/framework/README.md:99]: I encountered errors when including parameter `-n=2.`  This was removed from the Confluence instructions for now.  This should probably be removed from the README as well or replaced with a working command for parallel test execution.

File: uitests/framework/test_cases/conftest.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+ie = {
+    "os_version": "10",
+    "os": "Windows",
+    "browser": "ie",  # Edge  Chrome  IE
+    "browser_version": "11.0",  # 88.0  89.0    11.0
+    "resolution": "1920x1080",
+    "browserstack.sendKeys": "true",
+    "browserstack.debug": "true",
+    "build": buildName,  # Your tests will be organized within this build
+}
```
Comment (jasonburkert-ccpo) [inline uitests/framework/test_cases/conftest.py:39]: I suggest changing the `browser_version` to **latest** if possible.
If not possible, the Edge version should be changed to **89.0**.

**Review by jasonburkert-ccpo — CHANGES_REQUESTED** [review]
Browser versions in conftest.py should be set to latest so they update dynamically, or at a minimum change Edge to 89 so it's in sync with Chrome.

File `static/img/test.pdf` should be moved to `uitests/resources` or `tests/resources`.  The file should not be available in a hypothetical deployment at https://production.atat.dev/static/img/test.pdf

The parameter -n=2 should be removed from the pytest commands.  Myself and two other people have been unsuccessful attempting to include that parameter.

File: uitests/framework/README.md (hunk 9 above)
Comment (ehoward-ccpo) [reply uitests/framework/README.md:99]: Removed this section for now and will revisit at a later point.

File: uitests/framework/test_cases/conftest.py (hunk 10 above)
Comment (ehoward-ccpo) [reply uitests/framework/test_cases/conftest.py:39]: Updated the browser version to "latest" for all supported browsers

**Review by jasonburkert-ccpo — APPROVED** [review]
Peer reviewed.  All requested changes have been addressed.

**Comment by jasonburkert-ccpo** [comment]
Rebasing involved several merge conflicts to resolve.
`git pull --rebase origin develop && git push --force`

**Pull request merged**